### ISSUE TITLE
[error] can not append "object" as "Blob"

### ISSUE THREAD

**Issue by sylvanfrost** [form-data/form-data#529]
## environment
node - 15.7.0 (first version has blob as experimental)
form-data - ^4.0.0

## problem
1. I tried to append form data as Blob.
1. I need to append as real object not object as string.
1. If has other way to append form data as real object I will go that way but I don't know how. 

## why
My api develop by Hapi framework and validate by Joi. When api get a request Hapi will use Joi validate payload first (not my code), I need to validate some key in payload as my expected object
joi exmaple
```javascript
Joi.object({
  "a": Joi.object({
    "a": Joi.string(),
    "b": Joi.number().required()
  });
});
```

## remark
JavaScript in html file can append blob

## code
```javascript
const { Blob } = require("buffer");
const formData = require("form-data");

const objectHere = {
  "a": 1,
  "b": 2
}
const objectAsBlob = new Blob([JSON.stringify(objectHere, null, 0)], { type: "application/json" });

formdata.append("data", objectAsBlob); // error at this line
```

## error
```bash
> node ./index.js

Debugger attached.
Waiting for the debugger to disconnect...
C:\Users\username\poc\node_modules\delayed-stream\lib\delayed_stream.js:33
  source.on('error', function() {});
         ^

TypeError: source.on is not a function
    at Function.DelayedStream.create (C:\Users\username\poc\node_modules\delayed-stream\lib\delayed_stream.js:33:10)
    at FormData.CombinedStream.append (C:\Users\username\poc\node_modules\combined-stream\lib\combined_stream.js:45:37)
    at FormData.append (C:\Users\username\poc\node_modules\form-data\lib\form_data.js:75:3)
    at C:\Users\username\poc\index.js:48:24
    at Array.forEach (<anonymous>)
    at GenerateServerInjectionOptions (C:\Users\username\poc\index.js:44:23)
    at Object.<anonymous> (C:\Users\username\poc\index.js:67:14)
    at Module._compile (node:internal/modules/cjs/loader:1108:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1137:10)
    at Module.load (node:internal/modules/cjs/loader:973:32)
npm ERR! code 1
npm ERR! path C:\Users\username\poc
npm ERR! command failed
npm ERR! command C:\Windows\system32\cmd.exe /d /s /c node ./index.js

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\username\AppData\Local\npm-cache\_logs\2022-02-22T04_16_37_431Z-debug.log
Waiting for the debugger to disconnect...
```
## in html (work)
```html
<script>
  const xhr = new XMLHttpRequest();
  const formData = new FormData();
  
  const objectHere = {
    "a": 1,
    "b": 2
  }
  const objectAsBlob = new Blob([JSON.stringify(objectHere, null, 0)], { type: "application/json" });
  
  formData.append("data", objectAsBlob);
  
  xhr.open("POST", "http://localhost:8080/poc");
  xhr.send(formData); // can sent and receive 
</script>
```

**Comment by all-in-one-bravo** [comment]
You can try this. It works!
`formData.append('claimSubmission', JSON.stringify(request), { contentType: "application/json" });`

**Comment by anhvule** [comment]
You can try this. It works!
`formData.append('claimSubmission', JSON.stringify(request), { contentType: "application/json" });`

**Comment by sylvanfrost** [comment]
> FormData doesn't support Blob. Try this instead. It works!
> 
> ```
> formData.append('name', JSON.stringify(request), { contentType: "application/json" });
> ```
> 
> ```
> headers: {
>   "Authorization": "Bearer " + accessToken,
>   ...formData.getHeaders()
> }
> ```

thank you for answer, but I already try this way before like this...
```typescript
const generateRequest = () => {
  // somecode
  
  returnOptions.headers = { ...returnOptions.headers, ...FD.getHeaders(), ...options.headers };
  returnOptions.payload = FD.getBuffer();
  
  return returnOptions;
}
```
after this function, I send formData to my api but Joi cant validation `aa` field like object (joi throw aa field -> "This field must be an object")
```javascript
// my Joi Validation
const JoiValidation = Joi.Object({
  aa: joi.Array().items(Joi.Object({
    oa: Joi.string(),
    ob: Joi.string()
  }))
});
```

**Comment by Worik42** [comment]
same the problem. I have a blob pdf file. I need to insert it into the form data. NestJS, axios.

**Comment by ProdigyView** [comment]
Same problem, and according to the docs, FormData does support Blob: https://stackoverflow.com/questions/73465764/node-fetch-blobfromsync-and-blobfrom-typeerror-err-invalid-url-invalid-url

**Comment by ubudragon** [comment]
it's a nightmare with bytes \ blobs and form-data
it simply doesn't work at all

**Comment by niemal** [comment]
Yeah, I don't get it how it so broken. I tried grabbing the Blob's `stream()` and `getReader()` after that but still nothing.

**Comment by dodancs** [comment]
@anhvule 

> ```
> formData.append('name', JSON.stringify(request), { contentType: "application/json" });
> ```
> 
> ```
> headers: {
>   "Authorization": "Bearer " + accessToken,
>   ...formData.getHeaders()
> }
> ```

Thank you, this resolved my issue and it worked as expected!

**Comment by facadesss** [comment]
the same problem, i solved it with formdata-node package

**Comment by samabel98** [comment]
> FormData doesn't support Blob. Try this instead. It works!
> 
> ```
> formData.append('name', JSON.stringify(request), { contentType: "application/json" });
> ```
> 
> ```
> headers: {
>   "Authorization": "Bearer " + accessToken,
>   ...formData.getHeaders()
> }
> ```

but if i need it as blob to send it to multer , how can i do it ?

**Comment by CodeByBryant** [comment]
## TL;DR
- Node’s `form-data` library doesn’t accept a Blob object directly; use a Buffer/stream plus a filename and contentType.
- In the browser, native `FormData` does accept `Blob|File`.
- Multer only cares that it receives a file field in multipart/form-data; give it a filename and content type and it will parse it as a file.

## Node (using form-data)
If you need to send something as a blob to Multer from Node, convert it to a Buffer and append it with a filename and content type. That’s what Multer needs to treat it as a file.

Example: send JSON as a file part
```js
const FormData = require('form-data');
const axios = require('axios');

const formData = new FormData();

// Make a JSON Buffer (aka your “blob”)
const json = JSON.stringify({ hello: 'world' });
const buf = Buffer.from(json);

// Append as a file part with a filename and content type
formData.append('file', buf, {
  filename: 'request.json',
  contentType: 'application/json',
});

// You can still append simple text fields too
formData.append('meta', 'some meta value');

const headers = {
  Authorization: 'Bearer ' + accessToken,
  ...formData.getHeaders(),
};

await axios.post('https://your-server/upload', formData, { headers });
```

Multer side (Express):
```js
const multer = require('multer');
const upload = multer({ storage: multer.memoryStorage() });

app.post('/upload', upload.single('file'), (req, res) => {
  // req.file is your uploaded "blob"
  // req.file.buffer is a Buffer containing the JSON
  // req.file.mimetype is application/json
  // req.file.originalname is request.json
  res.json({ ok: true, size: req.file.size, mimetype: req.file.mimetype });
});
```

If you actually have binary data (e.g., an image), use its Buffer and a proper filename and contentType:
```js
formData.append('file', imageBuffer, {
  filename: 'photo.jpg',
  contentType: 'image/jpeg',
});
```

## Browser (native FormData)
In the browser you can append a Blob/File directly:
```js
const formData = new FormData();

const json = JSON.stringify({ hello: 'world' });
const blob = new Blob([json], { type: 'application/json' });

formData.append('file', blob, 'request.json');

await fetch('/upload', {
  method: 'POST',
  headers: { Authorization: 'Bearer ' + accessToken },
  body: formData,
});
```

Multer will parse it the same way as above.

## Why your current code “works” for fields but not blobs
This:
```js
formData.append('name', JSON.stringify(request), { contentType: 'application/json' });
```
creates a regular text field named `name`, not a file part, and Multer won’t put it under `req.file`/`req.files`. To make Multer treat it as a file, you must:
- use a binary value (Buffer/stream in Node, Blob/File in browser),
- provide a filename,
- set the content type.

**Comment by ljharb** [comment]
It'd be great to review a (non-AI-generated) PR to add Blob support to form-data, especially now that node has a built-in `Blob`.
